Disc 195/duration bug

jorntx opened the pull request:
- Fixed duration bug
- Now duration handles hours as well as minutes and seconds
- Moved duration handling to separate vue component

jesperlauridsen commented on src/components/records/components/Duration.vue:
```diff
@@ -0,0 +1,57 @@
+					console.log('No match found for the ISO duration format.');
+				}
+			} else {
+				console.log('No ISO duration provided.');
```
Logging that no ISO duration was found is nice - but useless for users who doesn't check the console. Perhaps it should read "Ukendt varighed" / "Unknown duration" or something similar if there was no duration found?

jorntx replied:
Good point and I agree. Actually I think we should suppress it all together because we will run into cases where we have start/end time but no duration. 

The gold plate solution would be to supply both start/end time to the duration component and if there is no precalculated duration we do the work ourself. If no duration AND no start or end time we write unknown duration.

jorntx replied:
I went for the gold plate solution see comments below...:)

jorntx commented:
So I went for the gold plated solution, fixed the bug and added the missing translation and did some hardening of the duration parser. The full commit log:
 

- Restructure the duration methods to handle translations
- Add skeleton for unit tests
- Bump and changing dependencies and jest/babel/eslint
- Change jest/babel config
- Add unit test for various cases of bogues duration and start/end dates
- Disable Legacy mode for i18n so it works with composition API mode
- Restructered and hardened duration methods to make tests pass
- Add duration translations
- Change typing in BrodcastRecord
- Fix missing support for $tc in hitcount and notification

jesperlauridsen commented on src/components/records/components/Duration.vue:
```diff
@@ -0,0 +1,104 @@
+	name: 'Duration',
+	props: {
+		isoDuration: {
+			type: String as PropType<string>,
```
What is the difference here, from just doing type:string? If I do type: String it still seems valid? What am I missing here?

jorntx replied:
Very good catch! - I have been messing with the typescript validator but ended up using an older version instead.

In the version I upgraded to I got an error when using only String - but this version does not really care so actually you are right - `String` would work perfectly fine and would be more concise. The only reason for going down this road would be if you were using Typescript to på more concise --> `String as PropType<Duration> `

jesperlauridsen commented on src/components/records/components/Duration.vue:
```diff
@@ -0,0 +1,104 @@
+		},
+	},
+
+	setup(props) {
```
You've build everything into the setup function here - is there a reason for not having the methods set up in the methods part instead?

jorntx replied:
The reason for building everything in the setup function is twofold here:

1. The composition API best practice is to have it in the setup function and then export the methods needed in the templates (Yes, I know we have inconsistency on this in our app :) )
2. In this case everything is handled on onMounted nothing is called from the template, so the methods naturally belong to the core of the component as "private" methods and this is setup() in the composition API. If it gets too messy visually methods can be moved out from setup() and just referenced/called in setup.

jesperlauridsen commented on src/components/records/components/Duration.vue:
```diff
@@ -0,0 +1,104 @@
+			const endDate = props.endDate;
+
+			if (isoDuration && isoDuration !== '') {
+				const match = isoDuration.match(/PT(\d+H)?(\d+M)?(\d+S)?/);
```
Regex man. It *looks* solid!

jorntx replied:
Stackoverflow for the win!

jesperlauridsen reviewed (approved):
I have a few comments, but they are mostly of a curious nature. Approved ! :1st_place_medal: Well done with the goldplating! A new level has been achieved with the unit tests too!

jorntx reviewed (commented):
Already merged - answers to comments in comments :)
